### PR TITLE
additional stringprep profiles

### PR DESCRIPTION
**Pull request by esessoms**
I was glad to find your work on stringprep.  I've added 3 more profiles (nameprep, nodeprep, and resourceprep) that are needed for XMPP.  Would you consider including these in your project?  Thanks!

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
+extern crate regex;
```
Comment (sfackler) [inline build.rs]: It'd probably be a good idea to run this once and check the generated code in. There's no real reason to go through the effort of doing this every time the crate's compiled.

File: src/tables.rs [hunk 2]
```diff
@@ -13,6 +55,31 @@ pub fn commonly_mapped_to_nothing(c: char) -> bool {
...
 }
 
+/// B.2 Mapping for case-folding used with NFKC.
+pub fn case_fold(s: &str) -> String {
+    let mut result = String::new();
+
+    // Each character either maps to a sequence of replacement characters,
+    // or is passed through as-is.
+    for c in s.chars() {
+        if let Some(entry) = table_lookup(RFC3454_B_2, c) {
```
Comment (sfackler) [inline src/tables.rs]: This could be delegating to `unicode_normalization`, right?

File: src/lib.rs [hunk 3]
```diff
@@ -14,15 +14,18 @@ use unicode_normalization::UnicodeNormalization;
 
 pub mod tables;
 
+/// Describes why a string failed stringprep normalization.
 #[derive(Debug)]
-enum ErrorCause {
+pub enum ErrorCause {
```
Comment (esessoms) [inline src/lib.rs]: That does make sense, but you expose `Error`:

https://github.com/sfackler/rust-stringprep/blob/master/src/lib.rs#L25

And keeping `ErrorCause` private gives no way to match it:

```
 --> tests/saslprep_tests.rs:8:13
  |
8 |         Err(Error(_)) => (),
  |             ^^^^^
  |             |
  |             did you mean `Error { /* fields */ }`?
  |             constructor is not visible here due to private fields
```

Your call, of course, but I think it needs to be fixed in one direction or the other.

File: build.rs (hunk 1 above)
Comment (esessoms) [inline build.rs]: Can do.

File: src/tables.rs (hunk 2 above)
Comment (esessoms) [inline src/tables.rs]: Maybe I'm blind, bud I don't see it:

https://unicode-rs.github.io/unicode-normalization/unicode_normalization/trait.UnicodeNormalization.html

Would have saved me a lot of work :) but I was unable to find any existing Rust implementation of case-folding.

**Comment by esessoms** [comment]
Pushed a new commit that gets rid of the build step.  I did *not* squash.

It occurs to me there's at least one good reason to keep the build step, which is... bugs.  Nobody really wants to maintain 1800 lines of lookup tables.

Right now you can have it either way, but if you're sure, I can go ahead and clean up the commit history.

File: src/lib.rs (hunk 3 above)
Comment (esessoms) [inline src/lib.rs]: NVM.  I understand.

**Comment by sfackler** [comment]
The maintenance would be the same either way - the lookup table file would be automatically generated and never manually edited.

**Comment by esessoms** [comment]
I don't really speak CircleCI, but I'm pretty sure "no jobs found" is not me.  Anyway, latest commit basically just archives the old build script, in case it is ever needed to regenerate lookup tables.  And, you're right, the build is *way* faster this way.

**Comment by sfackler** [comment]
I think Circle may be confused since I turned on cross-repo tests while the PR was already open. Looks good, thanks!

File: src/tables.rs (hunk 2 above)
Comment (sfackler) [inline src/tables.rs]: https://unicode-rs.github.io/unicode-normalization/unicode_normalization/trait.UnicodeNormalization.html#tymethod.nfkc this returns an iterator over the folded characters - it's used in the saslprep implementation already.

Comment (sfackler) [inline src/tables.rs]: Oh nevermind, sorry. I was thinking of section 4, not 3.2

**Comment by esessoms** [comment]
Thanks!

**Comment by sfackler** [comment]
Released 0.1.2 with these changes!

**Pull request merged**